pkg/statsd/handler_cloud.go

```diff
@@ -226,13 +305,21 @@ func (ch *CloudHandler) Run(ctx context.Context) {
 		logger:        ch.logger,
 	}
 
-	var wg wait.Group
-	defer wg.Wait() // Wait for lookupDispatcher to stop
+	// Stager will perform ordered, graceful shutdown. Stage by stage in reverse startup order.
+	stgr := stager.New()
+	defer stgr.Shutdown() // Wait for CloudProvider and lookupDispatcher to stop
+
+	stage := stgr.NextStage()
 
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel() // Tell lookupDispatcher to stop
+	// If the cloud provider implements runnable, it has things to do. It must start before anything that
+	// accesses it
+	if ch.cloud != nil {
+		if r, ok := ch.cloud.(gostatsd.Runner); ok {
+			r.Run(ctx) // synchronous - must fork inside if necessary
```

Comment:
By convention `Run()` of the `Runner` interface is a blocking method. If should be launched in a stage of stager.

Reply:
I get that it is convention but to allow waiting for the cache in a nice, cloud provider agnostic way, I thought tying in the cache wait would allow startup to happen in a nicer order. If you can think of a better way to achieve this I would love to change this because I agree on the convention.